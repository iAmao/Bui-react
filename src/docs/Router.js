import React from 'react';
import PropTypes from 'prop-types';


class Router extends React.Component {
    state = {
        location: {
            path: window.location.pathname,
        }
    };

    componentDidMount() {
        window.onpopstate = this.popState;
    }

    componentWillUnmount() {
        window.onpopstate = null;
    }

    popState = (event) => {
        this.setState({ location: { path: window.location.pathname } });
    }

    push = (path) => {
        this.setState({ location: { path } });
        window.history.pushState(window.history.state, '', path);
    }

    back = () => {
        window.history.back();
    }

    getChildContext() {
        return {
            location: {
                path: this.state.location.path,
                push: this.push,
                back: this.back
            }
        };
    }

    render() {
        return this.props.render({
            path: this.state.location.path,
            push: this.push,
            back: this.back
        });
    }
}

Router.childContextTypes = {
    location: PropTypes.shape({
        path: PropTypes.string,
        push: PropTypes.func,
        back: PropTypes.func
    })
};

Router.propTypes = {
    render: PropTypes.func.isRequired
}

export default Router;
