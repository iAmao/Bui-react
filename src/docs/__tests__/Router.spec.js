import React from 'react';
import renderer from 'react-test-renderer';

import { shallow } from 'enzyme';

import Router from '../Router';


describe('Doc: Router', () => {
    describe('Snapshot Test', () => {
        it('renders the component page', () => {
            const tree = renderer.create(
                <Router render={() => <div id="some" />} />
            ).toJSON();
            expect(tree).toMatchSnapshot();
        });
    });

    describe('Unit Test', () => {
        it('should pop history state', () => {
            const wrapper = shallow(<Router render={() => <div />} />);
            wrapper.instance().popState()
            expect(wrapper.instance().state).toEqual({ location: { path: '/' } })
        });

        it('should set a new path as state and push to new state', () => {
            const wrapper = shallow(<Router render={() => <div />} />);
            wrapper.instance().push('/button')
            expect(wrapper.instance().state).toEqual({ location: { path: '/button' } })
        });

        it('should go back to previous path', () => {
            window.history.pushState(null, '', '/item');
            const wrapper = shallow(<Router render={() => <div />} />);
            window.history.pushState(null, '', '/button');
            wrapper.instance().back()
            expect(wrapper.instance().state).toEqual({ location: { path: '/item' } })
        });
    });
});
