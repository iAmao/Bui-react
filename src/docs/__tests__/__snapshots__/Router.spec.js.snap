// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Doc: Router Snapshot Test renders the component page 1`] = `
<div
  id="some"
/>
`;
